File: 2022-06/2022-06-06/143. Reorder List/solution.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def _reverse(self, head: ListNode) -> ListNode:
        curr = head

        while curr.next:
            temp = curr.next
            curr.next = curr.next.next
            temp.next = head
            head = temp

        return head

    def _find_mid(self, head: ListNode) -> ListNode:
        slow = head
        fast = head.next

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        return slow

    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """

        if head is None or head.next is None:
            return

        slow = self._find_mid(head)

        l1 = head
        l2 = slow.next
        slow.next = None

        l2 = self._reverse(l2)

        while l1 and l2:
            temp = l2.next
            l2.next = l1.next
            l1.next = l2
            l1 = l1.next.next
            l2 = temp

File: 2022-06/2022-06-06/143. Reorder List/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func reorderList(head *ListNode)  {
	 slow := head
	 fast := head.Next

	 for fast != nil && fast.Next != nil {
		 fast = fast.Next.Next
		 slow = slow.Next
	 }

	 secondHalf := reverse(slow)
	 slow.Next = nil

	 fmt.Println(secondHalf)

	 cur := head
	 for cur != nil && secondHalf != nil {
		 next := cur.Next
		 secondHalfNext := secondHalf.Next

		 cur.Next = secondHalf
		 secondHalf.Next = next

		 cur = next
		 secondHalf = secondHalfNext
	 }

 }

 func reverse(head *ListNode) *ListNode {
	 cur := head
	 for cur.Next != nil {
		 temp := cur.Next
		 cur.Next = temp.Next
		 temp.Next = head
		 head = temp
	 }

	 return head
 }